File: python/phone-number/phone_number.py
import re

class PhoneNumber:
    def __init__(self, number):
        temp = "".join(re.findall(r'\d+', number))
        
        if len(temp) > 11:
            raise ValueError("Invalid number")
        
        if len(temp) == 11:
            if temp[:1] != "1":
                raise ValueError("Invalid number")
            else:
                temp = temp[1:]
        
        if int(temp[:1]) < 2 or int(temp[3:4]) < 2:
            raise ValueError("Invalid number")
        
        self.number = temp
        self.area_code = self.number[0:3]

    def pretty(self):
        return f"({self.area_code}) {self.number[3:6]}-{self.number[6:]}"

File: go/phone-number/phone_number.go
package phonenumber

import (
	"fmt"
	"strings"
	"unicode"
)

func Number(phoneNumber string) (string, error) {
	var sb strings.Builder
	var pos int
	for _, c := range phoneNumber {
		if !unicode.IsNumber(c) {
			continue
		}
		number := c - '0'
		// Ignore country code
		if pos == 0 && number == 1 {
			continue
		}
		if (pos == 0 || pos == 3) && number < 2 {
			return "", fmt.Errorf("invalid number")
		}
		sb.WriteRune(c)
		pos++
	}
	result := sb.String()

	if len(result) != 10 {
		return "", fmt.Errorf("invalid phone number length")
	}
	return result, nil
}

func AreaCode(phoneNumber string) (string, error) {
	parsed, err := Number(phoneNumber)
	if err != nil {
		return "", err
	}
	return parsed[:3], nil
}

func Format(phoneNumber string) (string, error) {
	parsed, err := Number(phoneNumber)
	if err != nil {
		return "", err
	}
	return fmt.Sprintf("(%s) %s-%s", parsed[:3], parsed[3:6], parsed[6:]), nil
}
